pkg/agent/host_configuration_linux.go

```diff
@@ -0,0 +1,135 @@
+			if link, err := netlink.LinkByName(i.hostGateway); err != nil {
+				klog.V(2).Infof("Not found host link for gateway %s, retry after 1s", i.hostGateway)
+				if _, ok := err.(netlink.LinkNotFoundError); ok {
+					time.Sleep(1 * time.Second)
+				} else {
+					return link, err
+				}
+			} else {
+				return link, nil
+			}
+		}
+		return nil, fmt.Errorf("link %s not found", i.hostGateway)
+	}()
+	if err != nil {
+		klog.Errorf("Failed to find host link for gateway %s: %v", i.hostGateway, err)
+		return err
+	}
+
+	// Set host gateway interface up
+	if err := netlink.LinkSetUp(link); err != nil {
+		klog.Errorf("Failed to set host link for %s up: %v", i.hostGateway, err)
+		return err
+	}
+
+	// Configure host gateway IP using the first address of node localSubnet
+	localSubnet := i.nodeConfig.PodCIDR
+	subnetID := localSubnet.IP.Mask(localSubnet.Mask)
+	gwIP := &net.IPNet{IP: ip.NextIP(subnetID), Mask: localSubnet.Mask}
+	gwAddr := &netlink.Addr{IPNet: gwIP, Label: ""}
+	gwMAC := link.Attrs().HardwareAddr
+	i.nodeConfig.GatewayConfig = &types.GatewayConfig{Name: i.hostGateway, IP: gwIP.IP, MAC: gwMAC}
+	gatewayIface.IP = gwIP.IP
+	gatewayIface.MAC = gwMAC
+
+	// Check IP address configuration on existing interface, return if already has target
+	// address
+	// We perform this check unconditionally, even if the OVS port did not exist when this
+	// function was called (i.e. portExists is false). Indeed, it may be possible for the Linux
+	// interface to exist even if the OVS bridge does not exist.
+	if addrs, err := netlink.AddrList(link, netlink.FAMILY_V4); err != nil {
+		klog.Errorf("Failed to query IPv4 address list for interface %s: %v", i.hostGateway, err)
+		return err
+	} else if addrs != nil {
+		for _, addr := range addrs {
+			klog.V(4).Infof("Found IPv4 address %s for interface %s", addr.IP.String(), i.hostGateway)
+			if addr.IP.Equal(gwAddr.IPNet.IP) {
+				klog.V(2).Infof("IPv4 address %s already assigned to interface %s", addr.IP.String(), i.hostGateway)
+				return nil
+			}
+		}
+	} else {
+		klog.V(2).Infof("Link %s has no configured IPv4 address", i.hostGateway)
+	}
+
+	klog.V(2).Infof("Adding address %v to gateway interface %s", gwAddr, i.hostGateway)
+	if err := netlink.AddrAdd(link, gwAddr); err != nil {
+		klog.Errorf("Failed to set gateway interface %s with address %v: %v", i.hostGateway, gwAddr, err)
+		return err
+	}
+
+	// send_redirects for the interface will be enabled if at least one of
+	// conf/{all,interface}/send_redirects is set to TRUE, so "all" and the
+	// interface must be disabled together.
+	// See https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt.
+	if err := disableICMPSendRedirects("all"); err != nil {
+		return err
+	}
+	if err := disableICMPSendRedirects(i.hostGateway); err != nil {
+		return err
+	}
+	return nil
+}
+
+// prepareHostNetworking returns immediately on Linux.
+func (i *Initializer) prepareHostNetworking() error {
```

Comment:
Do you think it makes sense to create a separate interface for OS specific funcs which have different implementations for different OSes, rather than reusing Initializer?
The benefit is reader can easily figure out the OS specific funcs. I am not very sure if it is right way in Go or not.
@tnqn

Reply:
So far I see a lot of such cases in K8s are done via _OS.go approach, the reason in my mind is:

- clean code, no need to have branch/options to switch implementation

-  bit size saving (even in interface way, we shouldn't and can't compile windows impl in linux bit, so there's going to be _OS.go anyway, then why it needs to have an option to select OS based implementation?)
- it can do more and has less requirement (the different part can be a function, a variable, a struct) while interface way requires well-designed interfaces and strongly consistent code on interface consumer side.

But I haven't experienced this before, I will take a look at more Go projects and suggest.

Reply:
Not sure if I got all your mean. But even we have an interface, we can still achieve what you said (no extra branch/option, the interface implementation can still be in _OS.go)? If the interface implementation is in the same package then almost no difference from convenience perspective?

Reply:
Yes, but having this interface or not is not related to OS implementation but whether we want to make it abstract. I'm not against abstracting interface, just prefer not calling platform specific functions with platform specific names on main path. 
For example, having a generic function `setupExternalNetworking` to call `i.iptablesClient.Initialize` in host_configuration_linux.go looks better to me than calling `i.iptablesClient.Initialize` directly with a dummy implementation on main path.

Reply:
Agreed on what you said, we should define generic interface/funcs rather than OS specific ones like iptablesXYZ.